src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala

```diff
@@ -235,6 +238,61 @@ class UserService(protected val userInfo: UserInfo, val dataSource: SlickDataSou
     }
   }
 
+  private def verifyNoSubmissions(userRef: RawlsUserRef, dataAccess: DataAccess): ReadAction[PerRequestMessage] = {
+    dataAccess.submissionQuery.findBySubmitter(userRef.userSubjectId.value).exists.result flatMap {
+      case true => DBIO.failed(new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, "Cannot remove a user with submissions")))
+      case _ => DBIO.successful(RequestComplete(StatusCodes.OK))
+    }
+  }
+
+  // can't just delete DB rows; also removes user from Google Groups
+  private def removeUserFromAllGroups(userRef: RawlsUserRef, dataAccess: DataAccess): ReadWriteAction[RawlsUser] = {
+    withUser(userRef, dataAccess) { user =>
+      dataAccess.rawlsGroupQuery.listGroupsForUser(userRef) flatMap { groupRefs =>
+        // Slick complains if it remains a Set
+        val removeActions = groupRefs.toSeq.map { groupRef =>
+          withGroup(groupRef, dataAccess) { group =>
+            updateGroupMembersInternal(group, Set(user), Set.empty, RemoveGroupMembersOp, dataAccess)
+          }
+        }
+
+        DBIO.sequence(removeActions) flatMap { removeResults =>
+          reduceErrorReports(removeResults) match {
+            case Some(errorReport) => DBIO.failed(new RawlsExceptionWithErrorReport(errorReport))
+            case None => DBIO.successful(user)
+          }
+        }
+      }
+    }
+  }
+
+  private def deleteUserFromDB(userRef: RawlsUserRef, dataAccess: DataAccess): ReadWriteAction[Int] = {
+    dataAccess.rawlsUserQuery.findUserBySubjectId(userRef.userSubjectId.value).delete flatMap {
+      case 1 => DBIO.successful(1)
+      case rowsDeleted =>
+        val error = new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.InternalServerError, s"Expected to delete 1 row from user table, but deleted $rowsDeleted"))
+        DBIO.failed(error)
+    }
+  }
+
+  def removeUser(userRef: RawlsUserRef): Future[PerRequestMessage] = {
+    val dbRemoval = dataSource.inTransaction { dataAccess =>
+      for {
+        _ <- verifyNoSubmissions(userRef, dataAccess)
```

Comment:
I think the only risk of things getting messed up is on the Google side of things. Since we don't have any sort of rollbacks for deleting proxy groups and removing users from groups (all_users, for example), it may put the user into a weird state of permissions. But that shouldn't put things into an irrecoverable state because they can still delete workspaces which contain submissions that they've run and they can hit the register endpoint again to recreate the now-missing Google groups